app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt

```diff
@@ -0,0 +1,610 @@
+     *
+     * @param anchorMiddleXCoord x-coordinate for the middle of the anchor.
+     * @param popupStartCoord x-coordinate for the popup start
+     * @param arrowIndicatorWidth Width of the arrow indicator.
+     */
+    @Composable
+    private fun computeIndicatorArrowStartCoord(
+        anchorMiddleXCoord: Pixels,
+        popupStartCoord: Pixels,
+        arrowIndicatorWidth: Pixels
+    ): Pixels {
+        val arrowIndicatorHalfWidth = arrowIndicatorWidth.value / 2
+
+        return if (LocalConfiguration.current.layoutDirection == View.LAYOUT_DIRECTION_LTR) {
+            val visiblePopupStartCoord = popupStartCoord.value + HORIZONTAL_PADDING.dp.toPx()
+            val arrowIndicatorStartCoord = anchorMiddleXCoord.value - arrowIndicatorHalfWidth
+
+            Pixels((visiblePopupStartCoord - arrowIndicatorStartCoord).absoluteValue)
+        } else {
+            val indicatorStartCoord = popupStartCoord.value - HORIZONTAL_PADDING.dp.toPx() -
+                anchorMiddleXCoord.value - arrowIndicatorHalfWidth
+
+            Pixels(indicatorStartCoord.absoluteValue)
+        }
+    }
+
+    /**
+     * Cleanup and remove the current popup from the screen.
+     * Clients are not automatically informed about this. Use a separate call to [onDismiss] if needed.
+     */
+    @VisibleForTesting
+    internal fun dismiss() {
+        anchor.removeOnAttachStateChangeListener(anchorDetachedListener)
+        GeckoScreenOrientation.getInstance().removeListener(orientationChangeListener)
+        disposeComposition()
+        ViewTreeLifecycleOwner.set(this, null)
+        ViewTreeSavedStateRegistryOwner.set(this, null)
+        windowManager.removeViewImmediate(this)
+    }
+
+    /**
+     * Create fullscreen translucent layout params.
+     * This will allow placing the visible popup anywhere on the screen.
+     */
+    @VisibleForTesting internal fun createLayoutParams(): WindowManager.LayoutParams =
+        WindowManager.LayoutParams().apply {
+            type = WindowManager.LayoutParams.TYPE_APPLICATION_PANEL
+            token = anchor.applicationWindowToken
+            width = WindowManager.LayoutParams.MATCH_PARENT
+            height = WindowManager.LayoutParams.MATCH_PARENT
+            format = PixelFormat.TRANSLUCENT
+            flags = WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN or
+                WindowManager.LayoutParams.FLAG_HARDWARE_ACCELERATED
+        }
+
+    /**
+     * Intended to allow querying the insets of the navigation bar.
+     * Value will be `0` except for when the screen is rotated by 90 degrees.
+     */
+    private fun getLeftInsets() = ViewCompat.getRootWindowInsets(anchor)
+        ?.getInsets(WindowInsetsCompat.Type.systemBars())?.left
+        ?: 0.coerceAtLeast(0)
+
+    @Px
+    private fun Dp.toPx(): Int {
+        return this.value
+            .dpToPx(anchor.resources.displayMetrics)
+            .roundToInt()
+    }
```

Comment:
why use this over `fun Dp.roundToPx()` in `Density`?

Reply:
This allows to just write `dpValue.toPx()` to get what I want
vs having to write
`LocalDensity.current.run { dpValue.roundToPx() }`

Maybe another something to extract for later usages.